tests/binary_agreement_mitm.rs

```diff
@@ -0,0 +1,440 @@
+        msg_type: MessageType::BVal,
+        msg_contents: Some(BoolFromState::AEstimated(false)),
+        msg_count: NODES_PER_GROUP * 2,
+    },
+    Stage {
+        source_groups: &[3],
+        dest_groups: &[0, 1],
+        msg_type: MessageType::BVal,
+        msg_contents: Some(BoolFromState::AEstimated(true)),
+        msg_count: NODES_PER_GROUP * 2,
+    },
+    // !! Not mentioned in the GitHub issue, but seems necessary.
+    // F sends Aux(_) to A, because nodes in A need 2f+1 Aux messages before they broadcasts their coins.
+    Stage {
+        source_groups: &[3],
+        dest_groups: &[0, 1],
+        msg_type: MessageType::Aux,
+        msg_contents: Some(BoolFromState::AEstimated(false)),
+        msg_count: NODES_PER_GROUP * 2,
+    },
+    // Now all nodes in A broadcast their threshold shares over the coin, so since |A|+|F|=2f+1, the adversary can construct the random coin value s.
+    Stage {
+        source_groups: &[0, 1],
+        dest_groups: &[3],
+        msg_type: MessageType::Coin,
+        msg_contents: None,
+        msg_count: NODES_PER_GROUP * 2,
+    },
+    // The nodes in F send BVAL(\neg s) to all the nodes in B, and all the BVAL(\neg s) messages from nodes in A are delivered to all nodes in B.
+    // Thus all the nodes in B broadcast AUX(\neg s).
+    Stage {
+        source_groups: &[0, 1, 3],
+        dest_groups: &[2],
+        msg_type: MessageType::BVal,
+        msg_contents: Some(BoolFromState::CoinValue(true)),
+        msg_count: (NODES_PER_GROUP * 2 + 1) * NODES_PER_GROUP,
+    },
+    // Deliver all AUX(\neg s) messages; there are 2f+1 of them, since either every node in A0 broadcast AUX(\neg s) or every node in A1 broadcast AUX(\neg s).
+    // Thus all nodes in B see 2f+1 AUX(\neg s) messages and get to the end of the round with bin_values=\neg s.
+    // Thus the nodes in B continue to the next round voting \neg s while the nodes in A continue to the next round voting s.
+    Stage {
+        source_groups: &[0, 1, 2, 3],
+        dest_groups: &[2],
+        msg_type: MessageType::Aux,
+        msg_contents: Some(BoolFromState::CoinValue(true)),
+        msg_count: (NODES_PER_GROUP + 1) * (NODES_PER_GROUP)
+            + (NODES_PER_GROUP * (NODES_PER_GROUP - 1)),
+    },
+    // At this point all messages from the round are delivered, and the process repeats.
+];
+
+/// An adversary for the reordering attack.
+/// Described here: https://github.com/amiller/HoneyBadgerBFT/issues/59#issue-310368284
+/// Excluding the first node, which is F,
+/// A0 is the first third of nodes, A1 is the second third, and the rest are B.
+struct ReorderingAdversary {
+    stage: usize,
+    stage_progress: usize,
+    sent_stage_messages: bool,
+    epoch: u32,
+    coin_state: CoinState<NodeId>,
+    /// The estimated value for nodes in A.
+    a_estimated: bool,
+    // TODO this is really hacky but there's no better way to get this value
+    netinfo_mutex: Arc<Mutex<Option<Arc<NetworkInfo<NodeId>>>>>,
```

Comment:
Getting the networkinfo into the adversary is a surprisingly difficult task. The closure that gets the netinfo must be created after the adversary is created, but cannot have a mutable reference to the adversary because it needs to be moved into the NetBuilder. I tried to extract the adversary from the NetBuilder in the closure, but with only a mutable reference to a box I can't downcast it to my adversary type. If I could convert it to an Any it'd work, but even making Adversary a super-trait requiring Any, I run into https://github.com/rust-lang/rust/issues/5665

Reply:
Another option is having the Adversary type as a type parameter of NetBuilder. This would make creating a network without an adversary more difficult, though we could just have a helper for that using NullAdversary. What do you think about that?

Reply:
Is `netinfo_mutex` only used to get the `invocation_id`? If so, don't use `NetworkInfo` inside the adversary. Pass only data that are needed to construct the nonce, or, in the current master, [`coin_id`](https://github.com/poanetwork/hbbft/blob/5f21c9a4e49acc9a95e72ba13ef1f594030c8f24/src/binary_agreement/binary_agreement.rs#L326).

Reply:
Yes, you are right. Another option would be to clone the `NetworkInfo` of the `NewNodeInfo`. It's read-only in `BinaryAgreement`, so it should be fine to clone it at the time of building the `VirtualNet`.

Reply:
That's what's currently done :)

Reply:
If you clone, you don't need the enclosing `Arc<Mutex<Option<_>>>` in the constructor argument of the adversary.